### ISSUE TITLE
StyleGuide - In Testing

### ISSUE THREAD

**Issue by mgechev** [angular/angular.io#1160]
To be discussed.

// cc @johnpapa @wardbell 

---

* Use [Jasmine](https://jasmine.github.io/) for unit testing.

  *Why?*: It is well supported and popular in the JavaScript community.

  *Why?*: It is widely popular in the Angular community. Jasmine is used for testing both AngularJS 1.x and Angular 2.

  *Why?*: It has up-to-date type definitions so you can have great development experience using TypeScript.

* Use Karma as test runner.

* Place your test files side-by-side with your client code.

  *Why?*: The tests are easier to find since they are always next to the code they are testing.

  *Why?*: When you update source code it is easier to go update the tests at the same time.

  *Why?*: The code may act as documentation of the tested component.

* Name the test file the following way `NAME_OF_THE_TESTED_UNIT.spec.EXT`:

  ```
  /* RECOMMENDED */
  about.ts
  about.spec.ts
  ```

* Place integration tests and tests that cover multiple code units into a separate `tests` directory in your project's root.

  **[Table of Contents](#table-of-contents)**

* Use protractor for End-to-End testing.

  *Why?*: It provides high-level API on top of the [Selenium WebDriver](http://www.seleniumhq.org/projects/webdriver/) and is supported by the Angular core team.

* Name the End-to-End test files using the following convention: `NAME_OF_THE_TESTED_UNIT.e2e.EXT`.

**Comment by mgechev** [comment]
I suggest to drop the styles of type:

> Use [technology placeholder]

The extension for e2e tests sounds natural, however, I think we can move the e2e tests to the lazy-loading boundaries they test.

**Comment by Foxandxss** [comment]
> Coupling the style guide with specific testing framework, test runner sounds like a bad idea.

Totally agree. Also, there are people who don't like Jasmine and use Mocha or something else with success.

**Comment by johnpapa** [comment]
Good points. Let's consider where we are with testing right now. Today we know how to make this work with karma as the runner and with jasmine as the framework. These are also what the CLI is working with.

I also have a Wallaby runner, which is cool BTW. 

I don;t think there is anything wrong with suggesting a set of tools for people to use. In fact, I think most folks just want to know a combination that works so they can grab them and go. With that in mind, I lean towards having the guide indicate we suggest Karma and Jasmine, and here is more info (link to examples) on how to do so. But we love Mocha too ( mocha is my preference ), and we will add more on Mocha later.

**Comment by Foxandxss** [comment]
Yes, I would simply tone down this one. Instead of "Use foo", "We recommend foo".

**Comment by johnpapa** [comment]
With the new terminology in the guide, this would be *Consider*
